developer_support_script/check_build_stability.py

```diff
@@ -1,45 +1,162 @@
+        """
+        builds = [build["number"] for build in self._json_test_metadata["builds"]]
+        self._raw_builds = builds
+        return builds
+
+    def get_current_and_complete_builds(self):
+        """
+        Return the current and complete builds from the metadata json
+
+        :return: the current and complete builds from the metadata json
+        """
+
+        current_build =  self._json_test_metadata["lastBuild"]["number"]
+        complete_build =  self._json_test_metadata["lastCompletedBuild"]["number"]
+
+        return current_build, complete_build
+
+    def remove_current_build_if_no_results(self):
+        """
+        Remove the current build from the list of builds if there
+        are no results for it
+
+        :return: the list of builds without the current build if there
+        """
+        if self._complete_build != self._current_build:
+            self._raw_builds.remove( self._current_build)
+        return self._raw_builds
+
+    def retrieve_test_data(self, test_num: int):
+        """
+        Retrieve test data for a given build number
+
+        :param test_num: the build number to retrieve test data for
+        :return: the test data json
+        """
+        print(f"Getting data for {test_num}")
+        try:
+            page = urllib.request.urlopen(self.system_tests_url.format(test_num))
+            test_json = json.loads(page.read())
+            return test_json
+        except http.IncompleteRead as err:
+            err_page = err.partial
+            print(f"Failed to get test data for {test_num}: {err_page}")
+            return err_page
+
+    @staticmethod
+    def add_failed_tests_to_counter(test_json: dict, failing_tests: Counter):
+        """
+        Add failed tests to a Counter object containing all failed tests
+
+        :param test_json: the test data json
+        :param failing_tests: the Counter object to add failed tests to
+        :return: the Counter object containing all failed tests
+        """
+        try:
+            for test in test_json["suites"]:
+                for case in test["cases"]:
+                    if case["status"] == "FAILED":
+                        failing_tests.update([case["className"] + "." + case["name"]])
+        except KeyError as err:
+            print(f"Failed to add failed tests to counter: {err}")
+        return failing_tests
+
 
-number_to_print = 15
+    def retrieve_failed_tests_from_builds(self):
+        """
+        Retreive tests data and return failed tests from builds
+        within a Counter object.
 
-failing_tests = Counter()
+        :return: a Counter object containing all failed tests
+        """
+        test_counter = Counter()
+        for test_num in self._proccessed_builds:
+            test_json = self.retrieve_test_data(test_num)
+            failing_tests = self.add_failed_tests_to_counter(test_json, test_counter)
+        self.failing_tests = failing_tests
```

Comment:
same as my comment on `get_all_builds` this both sets self.failing_tests, and returns failing tests, which it then procedes to set self.failing_tests with, it does not need to do both.